serialization/src/derive_support.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): L. Kuklinek
+
+//! Implementation tools used by code generated by derive macros
+#![doc(hidden)]
+
+pub use static_assertions as sa;
+
+use crate::Input;
+
+/// Input byte stream with a one-byte lookahead
+pub struct Peekable<'a, I> {
```

Comment:
Isn't it better to just make this a trait implemented for I instead of a struct that takes a reference THEN does the same thing to it? From what I'm seeing, the reference wrapper doesn't serve any purpose. Maybe I'm missing something.

Reply:
I don't think it is doable the way you describe. `Decode::decode` has to work for any type that implements `Input`. The new trait would have to be implemented for all types that implement `Input`.